### PR TITLE
Fix MHLO installed targets

### PR DESCRIPTION
**Pull request by yaochengji**
From [issue](https://github.com/onnx/onnx-mlir/issues/1655)

- add all the targets of high level dialect in mlir-hlo repo

PS: On my side I cannot reproduce `error while loading shared libraries: libChloOps.so.16git: cannot open shared object file` before this patch.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -150,7 +150,7 @@ add_subdirectory(third_party/pybind11)
 add_subdirectory(third_party/rapidcheck)
 
 if (ONNX_MLIR_ENABLE_MHLO)
-  add_subdirectory(third_party/mlir-hlo EXCLUDE_FROM_ALL)
```
Comment (sstamenova) [inline CMakeLists.txt:153]: Why is it necessary to remove this? My expectation is that this should be fine since we don't want to build it as part of the `all` target and only when needed by other targets.

File: src/Conversion/ONNXToMhlo/CMakeLists.txt [hunk 2]
```diff
@@ -2,11 +2,19 @@
...
-  )
+  MLIRMhloUtils
+  MhloDialect
+  MhloRegisterDialects
+  StablehloBase
+  StablehloOps
+  StablehloRegister
+  StablehloTypeInference
+
+  DESTINATION ${CMAKE_INSTALL_LIBDIR}
```
Comment (sstamenova) [inline src/Conversion/ONNXToMhlo/CMakeLists.txt:16]: Are these intended to be installed where onnx-mlir is installed? If so, you should follow the onnx-mlir pattern for choosing installation paths.

File: src/Conversion/ONNXToMhlo/CMakeLists.txt [hunk 3]
```diff
@@ -45,10 +53,10 @@ add_onnx_mlir_library(OMONNXToMhlo
   OMONNXRewrite
   OMONNXOps
   OMSupport
-  )
```
Comment (sstamenova) [inline src/Conversion/ONNXToMhlo/CMakeLists.txt:48]: Why? We generally follow the pattern of indenting the closing parenthesis, so the change should be the other way around - indent it everywhere.

**Review by sstamenova — CHANGES_REQUESTED** [review]

File: src/Conversion/ONNXToMhlo/CMakeLists.txt (hunk 3 above)
Comment (yaochengji) [reply src/Conversion/ONNXToMhlo/CMakeLists.txt]: This is what my auto-formatter do, fixed.

File: src/Conversion/ONNXToMhlo/CMakeLists.txt (hunk 2 above)
Comment (yaochengji) [reply src/Conversion/ONNXToMhlo/CMakeLists.txt]: Yes, changed to `lib` instead of `${CMAKE_INSTALL_LIBDIR}`

**Comment by gongsu832** [comment]
> PS: On my side I cannot reproduce `error while loading shared libraries: libChloOps.so.16git: cannot open shared object file` before this patch.

Sorry I'm a bit confused by this sentence. You mean you can use `onnxmlirczar/onnx-mlir` docker image to compile a model successfully? That shouldn't be possible. What did you actually try? Simply running the image should produce the error:
```
# docker run --rm -ti onnxmlirczar/onnx-mlir
onnx-mlir: error while loading shared libraries: libChloOps.so.16git: cannot open shared object file: No such file or directory
```

**Comment by yaochengji** [comment]
> You mean you can use `onnxmlirczar/onnx-mlir` docker image to compile a model successfully?

I used the docker file of onnx-mlir to build my own docker image, but I think it's similar to `onnxmlirczar/onnx-mlir`.
My error is the same as the [issue](https://github.com/onnx/onnx-mlir/issues/1655) submmiter's , not `onnx-mlir: error while loading shared libraries: libChloOps.so.16git: cannot open shared object file: No such file or directory`.

```
onnx-mlir --EmitLib AlexNet.onnx 
error: reduction dimension must be the same
error: Failed to scan onnx.MatMul parameters successfully
error: shape inference failed
```

**Comment by gongsu832** [comment]
> > You mean you can use `onnxmlirczar/onnx-mlir` docker image to compile a model successfully?
> 
> I used the docker file of onnx-mlir to build my own docker image, but I think it's similar to `onnxmlirczar/onnx-mlir`. My error is the same as the [issue](https://github.com/onnx/onnx-mlir/issues/1655) submmiter's , not `onnx-mlir: error while loading shared libraries: libChloOps.so.16git: cannot open shared object file: No such file or directory`.
> 
> ```
> onnx-mlir --EmitLib AlexNet.onnx 
> error: reduction dimension must be the same
> error: Failed to scan onnx.MatMul parameters successfully
> error: shape inference failed
> ```

Then I'm guessing your docker image either managed to install the missing libChloOps.so.16git (and potentially others) or you kept the source tree in your image.

**Comment by yaochengji** [comment]
> Then I'm guessing your docker image either managed to install the missing libChloOps.so.16git (and potentially others) or you kept the source tree in your image.

I tried with a clean docker image, but still could not reproduce it. And I tried to test it directly by building a docker image with `docker/Dockerfile.onnx-mlir` but I don't know how to specify the docker file's parameter.

@gongsu832 could you help test it on your side?

**Comment by gongsu832** [comment]
> > Then I'm guessing your docker image either managed to install the missing libChloOps.so.16git (and potentially others) or you kept the source tree in your image.
> 
> I tried with a clean docker image, but still could not reproduce it. And I tried to test it directly by building a docker image with `docker/Dockerfile.onnx-mlir` but I don't know how to specify the docker file's parameter.
> 
> @gongsu832 could you help test it on your side?

Why don't you just `docker pull onnxmlirczar/onnx-mlir`?

**Comment by yaochengji** [comment]
> Why don't you just `docker pull onnxmlirczar/onnx-mlir`?

Yes, this can reproduce. But the problem is that I cannot start the docker container. (I need to build onnx-mlir in the docker container)

**Comment by gongsu832** [comment]
> > Why don't you just `docker pull onnxmlirczar/onnx-mlir`?
> 
> Yes, this can reproduce. But the problem is that I cannot start the docker container. (I need to build onnx-mlir in the docker container)

Another thing to check, in your own docker image, did you build llvm-project with `-DBUILD_SHARED_LIBS=ON`? By default, llvm-project builds everything statically and then onnx-mlir will also build everything statically, then you won't see the problem since all the libraries will have been embedded into the onnx-mlir binary. You can check the size of your onnx-mlir binary. If it's a few hundred MBs, then you are building statically.

I'm thinking of adding a build option for the user image so that the build tree can be kept for debugging purpose. But before that, you can do the equivalent by:

- build a shared llvm-project image:
   `docker build --build-arg BUILD_SHARED_LIBS=ON -t llvm-project-shared -f docker/Dockerfile.llvm-project`
- build a shared onnx-mlir image:
   `docker build --build-arg BASE_IMAGE=llvm-project-shared -t onnx-mlir-shared -f docker/Dockerfile.onnx-mlir`
   (but remember to remove the line in the dockerfile `&& rm -rf /tmp/* /usr/bin/python llvm-project onnx-mlir`)

**Comment by gongsu832** [comment]
> ~: docker run onnxmlirczar/onnx-mlir bash
> onnx-mlir: error while loading shared libraries: libChloOps.so.16git: cannot open shared object file: No such f

The correct command is
```
docker run --rm -ti --entrypoint=bash onnxmlirczar/onnx-mlir
```
But this won't help too much since all the build sources have been removed so you need to rebuild anyway.

**Comment by yaochengji** [comment]
> But this won't help too much since all the build sources have been removed so you need to rebuild anyway.

I started a new dokcer container by `docker run --rm -ti --entrypoint=bash onnxmlirczar/onnx-mlir-dev`. And basically I followed this [doc](https://github.com/onnx/onnx-mlir/blob/main/docs/BuildOnLinuxOSX.md) to build llvm-project and onnx-mlir ( use`cmake --build . --target install` instead and `rm -rf onnx-mlir/build/*` after building). But still I cannot reproduce the `libChloOps.so` error.

**Comment by gongsu832** [comment]
> > But this won't help too much since all the build sources have been removed so you need to rebuild anyway.
> 
> I started a new dokcer container by `docker run --rm -ti --entrypoint=bash onnxmlirczar/onnx-mlir-dev`. And basically I followed this [doc](https://github.com/onnx/onnx-mlir/blob/main/docs/BuildOnLinuxOSX.md) to build llvm-project and onnx-mlir ( use`cmake --build . --target install` instead and `rm -rf onnx-mlir/build/*` after building). But still I cannot reproduce the `libChloOps.so` error.

As I mentioned above, you need to build everything **shared** to produce the problem. `onnxmlirczar/onnx-mlir-dev` builds everything static, and the doc you followed also builds everything static.

BTW, for `onnxmlirczar/onnx-mlir-dev` image, you don't need to specify `--entrypoint=bash`. But as I said, it builds everything in static. So you will have to rebuild both llvm-project and onnx-mlir in shared anyway, just like what I told you above if you are going to build your own shared image.

File: src/Conversion/ONNXToMhlo/CMakeLists.txt [hunk 4]
```diff
@@ -52,3 +64,7 @@ add_onnx_mlir_library(OMONNXToMhlo
 target_link_libraries(MhloDialect PUBLIC
   MLIRSparseTensorDialect
   )
+
+target_link_libraries(StablehloTypeInference PUBLIC
```
Comment (sstamenova) [inline src/Conversion/ONNXToMhlo/CMakeLists.txt:68]: Why is this needed? The two libs are both linked into the target above

**Comment by yaochengji** [comment]
@gongsu832 Thanks for the reminding of the static/shared diffrence, I rebuilt llvm-project and onnx-mlir in a shared way. Now I can reproduce the `libChloOps.so` error and already tested locally.

File: src/Conversion/ONNXToMhlo/CMakeLists.txt (hunk 4 above)
Comment (yaochengji) [reply src/Conversion/ONNXToMhlo/CMakeLists.txt:68]: But `StablehloTypeInference` links into `StablehloBase`, which is missing in `mlir-hlo` repo.

**Comment by gongsu832** [comment]
@jenkins-droid test this please

File: src/Conversion/ONNXToMhlo/CMakeLists.txt (hunk 4 above)
Comment (sstamenova) [reply src/Conversion/ONNXToMhlo/CMakeLists.txt:68]: Have you let the mhlo folks know, so that it can be fixed there as well?

**Review by sstamenova — APPROVED** [review]
LGTM, but please make sure that the mhlo dependency issues gets fixed in mhlo as well - either by letting them know or sending a PR.

File: src/Conversion/ONNXToMhlo/CMakeLists.txt (hunk 4 above)
Comment (yaochengji) [reply src/Conversion/ONNXToMhlo/CMakeLists.txt:68]: It is added. https://github.com/openxla/stablehlo/pull/456.

BTW, I find that the additional linking relationship below is already added in mlir-hlo repo.
```
target_link_libraries(MhloDialect PUBLIC
  MLIRSparseTensorDialect
  )
```

**Comment by yaochengji** [comment]
@gongsu832  @sstamenova , can this PR be merged now? I will delete the additional LINK_LIBS later.
